platforms/chibios/drivers/serial_usart.c

```diff
@@ -64,36 +57,75 @@ static inline void usart_clear(void) {
     }
 }
 
+#elif HAL_USE_SIO
+
+void clear_rx_evt_cb(SIODriver* siop) {
+    osalSysLockFromISR();
+    /* If errors occured during transactions this callback is invoked. We just
+     * clear the error sources and move on. We rely on the fact that we check
+     * for the success of the transaction by comparing the received/send bytes
+     * with the actual received/send bytes in the send/receive functions. */
+    sioGetAndClearEventsI(serial_driver);
+    osalSysUnlockFromISR();
+}
+
+static const SIOOperation serial_usart_operation = {.rx_cb = NULL, .rx_idle_cb = NULL, .tx_cb = NULL, .tx_end_cb = NULL, .rx_evt_cb = &clear_rx_evt_cb};
+
 /**
- * @brief Blocking send of buffer with timeout.
- *
- * @return true Send success.
- * @return false Send failed.
+ * @brief SIO Driver startup routine.
  */
-static inline bool send(const uint8_t* source, const size_t size) {
-    bool success = (size_t)sdWriteTimeout(serial_driver, source, size, TIME_MS2I(SERIAL_USART_TIMEOUT)) == size;
+static inline void usart_driver_start(void) {
+    sioStart(serial_driver, &serial_config);
+    sioStartOperation(serial_driver, &serial_usart_operation);
+}
+
+inline void driver_clear(void) {
+    osalSysLock();
+    while (!sioIsRXEmptyX(serial_driver)) {
+        msg_t dump = sioGetX(serial_driver);
+        (void)dump;
+    }
+    osalSysUnlock();
+}
+
+#else
+
+#    error Either the SERIAL or SIO driver has to be activated to use the usart driver for split keyboards.
+
+#endif
+
+inline bool send(const uint8_t* source, const size_t size) {
+    bool success = (size_t)chnWriteTimeout(serial_driver, source, size, TIME_MS2I(SERIAL_USART_TIMEOUT)) == size;
 
 #if !defined(SERIAL_USART_FULL_DUPLEX)
-    if (success) {
+    if (likely(success)) {
         /* Half duplex fills the input queue with the data we wrote - just throw it away.
-           Under the right circumstances (e.g. bad cables paired with high baud rates)
-           less bytes can be present in the input queue, therefore a timeout is needed. */
-        uint8_t dump[size];
-        return receive(dump, size);
+         * Under the right circumstances (e.g. bad cables paired with high baud rates)
+         * less bytes can be present in the input queue, therefore a timeout is needed. */
+        uint8_t dump[64];
+        size_t  bytes_left = size;
+
+        while (unlikely(bytes_left > 64)) {
+            if (unlikely(!receive(dump, 64))) {
```

Comment:
Will need to `#include "util.h"` if it's not already present.
```suggestion
            if (unlikely(!receive(dump, MIN(bytes_left, 64)))) {
```
Prevents reading extra if the slave just happens to respond within `SERIAL_USART_TIMEOUT`.

Reply:
I don't think that change is correct? This branch will only be executed if `bytes_left  > 64` so we can safely always read 64 byte.